### PR TITLE
Allows to choose SSL context for SMTP provider

### PR DESCRIPTION
**Pull request by potiuk**
This change add two options to choose from when SSL SMTP connection is created:

* default - for balance between compatibility and security
* none - in case compatibility with existing infrastructure is   preferred

The fallback is:

* The Airflow "email", "ssl_context"
* "default"

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/smtp/CHANGELOG.rst [hunk 1]
```diff
@@ -27,6 +27,18 @@
 Changelog
 ---------
```
Comment (pankajkoti) [inline airflow/providers/smtp/CHANGELOG.rst:29]: Do we need to add here 1.3.0 or is that taken care of later during the release?

I guess this has to go in a minor release as we're adding a config param?

File: airflow/providers/smtp/hooks/smtp.py [hunk 2]
```diff
@@ -109,6 +109,24 @@ def _build_client(self) -> smtplib.SMTP_SSL | smtplib.SMTP:
...
         smtp_kwargs["timeout"] = self.timeout
 
+        if self.use_ssl:
+            from airflow.configuration import conf
+
+            ssl_context_string = conf.get("smtp_provider", "SSL_CONTEXT", fallback=None)
+            if ssl_context_string is None:
+                ssl_context_string = conf.get("email", "SSL_CONTEXT", fallback=None)
+            if ssl_context_string is None:
+                ssl_context_string = "default"
```
Comment (pankajkoti) [inline airflow/providers/smtp/hooks/smtp.py:119]:
```suggestion
            if ssl_context_string is None:
                ssl_context_string = conf.get("email", "SSL_CONTEXT", fallback="default")
```

At this point can we fallback to `default`? or we would like to have an explicit check later? or the fallback has some disadvantages? just curious :)

File: airflow/providers/smtp/hooks/smtp.py [hunk 3]
```diff
@@ -109,6 +109,24 @@ def _build_client(self) -> smtplib.SMTP_SSL | smtplib.SMTP:
...
+            if ssl_context_string == "default":
+                ssl_context = ssl.create_default_context()
+            elif ssl_context_string == "none":
+                ssl_context = None
+            else:
+                raise RuntimeError(
+                    f"The email.ssl_context configuration variable must "
+                    f"be set to 'default' or 'none' and is '{ssl_context_string}."
```
Comment (pankajkoti) [inline airflow/providers/smtp/hooks/smtp.py]:
```suggestion
                    f"be set to 'default' or 'none' and is '{ssl_context_string}'."
```

File: airflow/providers/smtp/provider.yaml [hunk 4]
```diff
@@ -54,3 +54,27 @@ connection-types:
...
+          are fixing your certificate configuration. This can be typically done by upgrading to newer
+          version of the operating system you run Airflow components on,by upgrading/refreshing proper
+          certificates in the OS or by updating certificates for your mail servers.
+
+          If you do not set this option explicitly, it will use Airflow "email.ssl_context" configuration,
+          but if this configuration is not present, it will use "default" value.
+        type: string
+        version_added: 1.3.0
+        example: "default"
+        default: ~
```
Comment (pankajkoti) [inline airflow/providers/smtp/provider.yaml:80]: Why do we not wish to set default: "default" here? Looking at the description above looks like we do intend to set a default of "default".

File: airflow/providers/smtp/CHANGELOG.rst (hunk 1 above)
Comment (potiuk) [reply airflow/providers/smtp/CHANGELOG.rst:29]: Recently @eladkal asked not to - it will be added when we decide on releasing (release manager decides on the actual bumpt).

File: airflow/providers/smtp/hooks/smtp.py (hunk 2 above)
Comment (potiuk) [reply airflow/providers/smtp/hooks/smtp.py:119]: We do fallback to default. The complexity is that we really want to have an "smtp-provider" default separate from the "email" one. This is why we do not have "built-in" default for the "smtp_provider" one. If we have  a default there, we cannot really use "email" as a fallback if it is set in Airflow.

File: airflow/providers/smtp/provider.yaml (hunk 4 above)
Comment (potiuk) [reply airflow/providers/smtp/provider.yaml:80]: This is sa, ereason as above - this is a chain of config: 

1) check if "smtp_provider" / "ssl_context" is set - if yes , we use it., if not we go to point 2.
2) check if "emai" /  "ssl_context" is set - if yes - we use it, if not we fallback to "default" 

So we cannot set default on "smtp_provider" one, because then we will retrieve default in point 1 and we cannot check point 2 after.


The complexity here is that we have to handle several cases: 

1) what happens if we have Airflow 2.6 and no "emai"/ssl_context
2) what happens if ewe hav Airflow 2.7 and someone set "email" / "ssl_context" = "none" and did not set "smtp_provider" / "ssl_context"

So in other words - our direct fallback is whatever the value of "email" / "ssl_context" is and only if it is not set we use "default"

Comment (potiuk) [reply airflow/providers/smtp/provider.yaml:80]: I added a complete set of unit tests showing it :)

Comment (pankajkoti) [reply airflow/providers/smtp/provider.yaml:80]: understood 🙂 

Fail-safe 🚀

**Comment by potiuk** [comment]
Also for @hussein-awala -> I am not sure if this approach is best and how it will work with https://github.com/apache/airflow/pull/30531 - I just figured that we have to use "smtp_provider" section rather than "smtp" from the core - and I think we should follow similar approach like here - where "smtp" will go to "pre_2_7_0" defaults, but the  provider will use the "smtp_provider" section and fall back to the "smtp" one, similarly as I am falling back now to "email" / "ssl_context" - would love to hear from you on that one. 

As explained above, I am trying to make chain of defaults this way and added unit tests covering the behaviour

File: airflow/providers/smtp/provider.yaml (hunk 4 above)
Comment (potiuk) [reply airflow/providers/smtp/provider.yaml:80]: Back-compat engineering as recently @dstandish named it.

File: airflow/providers/smtp/CHANGELOG.rst [hunk 5]
```diff
@@ -27,6 +27,18 @@
...
 
+In case of SMTP SSL connection, the default context now uses "default" context
+
+The "default" context is Python's ``default_ssl_contest`` instead of previously used "none". The
```
Comment (ephraimbuddy) [inline airflow/providers/smtp/CHANGELOG.rst]:
```suggestion
The "default" context is Python's ``default_ssl_context`` instead of previously used "none". The
```

**Review by ephraimbuddy — APPROVED** [review]

**Review by pankajkoti — APPROVED** [review]

**Review by utkarsharma2 — APPROVED** [review]

**Comment by potiuk** [comment]
@hussein-awala -> would you like to comment on that one, re #30531 - or you are ok for now. I'd love to merge that one and implement IMAP change as well so that we can release new providers today :)

**Comment by potiuk** [comment]
Let me merge that one now, I think we can always change it in the future.

**Comment by hussein-awala** [comment]
I'm late to the party.

> I just figured that we have to use "smtp_provider" section rather than "smtp" from the core - and I think we should follow similar approach like here - where "smtp" will go to "pre_2_7_0" defaults, but the provider will use the "smtp_provider" section and fall back to the "smtp" one, similarly as I am falling back now to "email" / "ssl_context" - would love to hear from you on that one.

For smtp_provider, I think it would be better to use the connection extras to configure ssl_context (as we do with the other configurations), instead of using Airlfow config (or even provider config).
https://github.com/potiuk/airflow/blob/ca2f3013bcb123c4b3973a5b85de77094bf2c459/airflow/providers/smtp/hooks/smtp.py#L321-L355

Since we cannot use the operator/hook without a connection, IMO it's better if we provide a single way to configure them.

WDYT? could we move this conf to the connection before releasing the provider?
